otherlibs/systhreads/Makefile

```diff
@@ -14,95 +14,129 @@
 
-allopt: libthreadsnat.a threads.cmxa
+THREADS_BCOBJS := $(THREADS_SOURCES:.ml=.cmo)
+THREADS_NCOBJS := $(THREADS_SOURCES:.ml=.cmx)
 
-libthreads.a: $(BYTECODE_C_OBJS)
-	$(MKLIB) -o threads $(BYTECODE_C_OBJS) $(PTHREAD_LINK)
+MLIFILES=thread.mli mutex.mli condition.mli event.mli threadUnix.mli
+CMIFILES=$(MLIFILES:.mli=.cmi)
 
-st_stubs_b.o: st_stubs.c st_posix.h
-	$(BYTECC) -I../../byterun $(BYTECCCOMPOPTS) $(SHAREDCCCOMPOPTS) \
-	   -c st_stubs.c
-	mv st_stubs.o st_stubs_b.o
+all: lib$(LIBNAME).$(A) $(LIBNAME).cma $(CMIFILES)
 
+allopt: lib$(LIBNAME).$(A) $(LIBNAME).cmxa $(LIBNAME).cmxs $(CMIFILES)
+
+lib$(LIBNAME).$(A): $(BYTECODE_C_OBJS)
+	$(MKLIB) -o $(LIBNAME) $(BYTECODE_C_OBJS) $(PTHREAD_LINK)
+
+lib$(LIBNAME)nat.$(A): $(NATIVECODE_C_OBJS)
+ifeq "$(UNIX_OR_WIN32)" "unix"
 # Dynamic linking with -lpthread is risky on many platforms, so
 # do not create a shared object for libthreadsnat.
-libthreadsnat.a: $(NATIVECODE_C_OBJS)
-	$(AR) rc libthreadsnat.a $(NATIVECODE_C_OBJS)
-
-st_stubs_n.o: st_stubs.c st_posix.h
-	$(NATIVECC) -I../../asmrun -I../../byterun $(NATIVECCCOMPOPTS) \
-	            $(SHAREDCCCOMPOPTS) -DNATIVE_CODE -DTARGET_$(ARCH) \
-	            -DMODEL_$(MODEL) -DSYS_$(SYSTEM) -c st_stubs.c
-	mv st_stubs.o st_stubs_n.o
+	$(AR) rc $@ $^
+else # Windows
+	$(MKLIB) -o $(LIBNAME)nat $^ $(LDOPTS)
+endif
 
-threads.cma: $(THREAD_OBJS)
-	$(MKLIB) -ocamlc '$(CAMLC)' -o threads $(THREAD_OBJS) \
-	  -cclib -lunix $(PTHREAD_CAML_LINK)
+$(LIBNAME).cma: $(THREADS_BCOBJS)
+ifeq "$(UNIX_OR_WIN32)" "unix"
+	$(MKLIB) -o $(LIBNAME) -ocamlc '$(CAMLC)' -cclib -lunix $(PTHREAD_CAML_LINK) $^
+else # Windows
+	$(MKLIB) -o $(LIBNAME) -ocamlc "$(CAMLC)" -linkall $^
+endif
 
 # See remark above: force static linking of libthreadsnat.a
-threads.cmxa: $(THREAD_OBJS:.cmo=.cmx)
-	$(CAMLOPT) -a -o threads.cmxa $(THREAD_OBJS:.cmo=.cmx) \
-	  -cclib -lthreadsnat $(PTHREAD_CAML_LINK)
+$(LIBNAME).cmxa: $(THREADS_NCOBJS)
+ifeq "$(UNIX_OR_WIN32)" "unix"
+	$(CAMLOPT) -a -o $@ $^ -cclib -lthreadsnat $(PTHREAD_CAML_LINK)
+else # Windows
+	$(MKLIB) -o $(LIBNAME)nat -ocamlopt "$(CAMLOPT)" -linkall $^ $(LINKOPTS)
+	mv $(LIBNAME)nat.cmxa $(LIBNAME).cmxa
+	mv $(LIBNAME)nat.$(A) $(LIBNAME).$(A)
+endif
 
-# Note: I removed "-cclib -lunix" from the line above.
+# Note: I removed "-cclib -lunix" from the Unix line above.
 # Indeed, if we link threads.cmxa, then we must also link unix.cmxa,
 # which itself will pass -lunix to the C linker.  It seems more
 # modular to me this way. -- Alain
 
+$(LIBNAME).cmxs: $(LIBNAME).cmxa lib$(LIBNAME)nat.$(A)
+	$(CAMLOPT) -shared -o $@ -I . $< -linkall
+
+st_stubs_b.$(O): st_stubs.c $(HEADER)
+	$(BYTECC) -I../../byterun $(BYTECCCOMPOPTS) $(SHAREDCCCOMPOPTS) -c $<
+	mv st_stubs.$(O) $@
+
+st_stubs_n.$(O): st_stubs.c $(HEADER)
+	$(NATIVECC) -I../../asmrun -I../../byterun $(NATIVECCCOMPOPTS) \
+	  $(SHAREDCCCOMPOPTS) -DNATIVE_CODE -DTARGET_$(ARCH) \
+	  -DMODEL_$(MODEL) -DSYS_$(SYSTEM) -c st_stubs.c
```

Comment:
I'd suggest using `$<` as in the rule above.  And also `$(ROOTDIR)` instead of  `../..`.

Wouldn't it be better to copy the source file before compiling, instead of moving the resulting object file?  I'm thinking about parallel make.